fish-rust/src/env_dispatch.rs

```diff
@@ -0,0 +1,796 @@
+        table.add_anon(L!("fish_function_path"), handle_function_path_change);
+        table.add_anon(L!("fish_read_limit"), handle_read_limit_change);
+        table.add_anon(L!("fish_history"), handle_fish_history_change);
+        table.add_anon(
+            L!("fish_autosuggestion_enabled"),
+            handle_autosuggestion_change,
+        );
+        table.add_anon(
+            L!("fish_use_posix_spawn"),
+            handle_fish_use_posix_spawn_change,
+        );
+        table.add_anon(L!("fish_trace"), handle_fish_trace);
+        table.add_anon(
+            L!("fish_cursor_selection_mode"),
+            handle_fish_cursor_selection_mode_change,
+        );
+
+        table
+    });
+
+type NamedEnvCallback = fn(name: &wstr, env: &dyn Environment);
+type AnonEnvCallback = fn(env: &dyn Environment);
+
+#[derive(Default)]
+struct VarDispatchTable {
+    named_table: HashMap<&'static wstr, NamedEnvCallback>,
+    anon_table: HashMap<&'static wstr, AnonEnvCallback>,
+}
+
+// TODO: Delete this after input_common is ported (and pass the input_function function directly).
+fn update_wait_on_escape_ms(vars: &dyn Environment) {
+    let fish_escape_delay_ms = vars.get_unless_empty(L!("fish_escape_delay_ms"));
+    let var = crate::env::environment::env_var_to_ffi(fish_escape_delay_ms);
+    crate::ffi::update_wait_on_escape_ms_ffi(var);
+}
+
+impl VarDispatchTable {
+    fn observes_var(&self, name: &wstr) -> bool {
+        self.named_table.contains_key(name) || self.anon_table.contains_key(name)
+    }
+
+    /// Add a callback for the variable `name`. We must not already be observing this variable.
+    pub fn add(&mut self, name: &'static wstr, callback: NamedEnvCallback) {
+        let prev = self.named_table.insert(name, callback);
+        assert!(prev.is_none(), "Already observing {}", name);
+    }
+
+    /// Add an callback for the variable `name`. We must not already be observing this variable.
+    pub fn add_anon(&mut self, name: &'static wstr, callback: AnonEnvCallback) {
+        let prev = self.anon_table.insert(name, callback);
+        assert!(prev.is_none(), "Already observing {}", name);
+    }
+
+    pub fn dispatch(&self, key: &wstr, vars: &EnvStack) {
+        if let Some(named) = self.named_table.get(key) {
+            (named)(key, vars);
+        }
+        if let Some(anon) = self.anon_table.get(key) {
+            (anon)(vars);
+        }
+    }
+}
+
+fn handle_timezone(env_var_name: &wstr, vars: &dyn Environment) {
+    let var = vars.get_unless_empty(env_var_name).map(|v| v.as_string());
+    FLOGF!(
+        env_dispatch,
+        "handle_timezone() current timezone var: |",
+        env_var_name,
+        "| => |",
+        var.as_ref()
+            .map(|v| v.as_utfstr())
+            .unwrap_or(L!("MISSING/EMPTY")),
+        "|"
+    );
+    let name = env_var_name.to_string();
+    if let Some(var) = var {
+        setenv_lock(&name, &var.to_string(), true);
```

Comment:
I don't think we can translate wcs2zstring(env_var_name) as env_var_name.to_string() because the latter does not know about our encoding scheme for codepoints that lie outside the valid Unicode ranges. 
We should use `wcs2zstring` which returns the CString we need.
Same for var.to_string().

Reply:
It's an exported environment variable, presumably that is intended for use outside of fish itself and therefore shouldn't have private use codepoints? I think it would be a mistake to make those externally visible anywhere, no?

Reply:
if we are given `set var \xff`, that's a one-byte string that's not a valid UTF-8 sequence.
We use `encode_byte_to_char()` to convert it to a valid wide string, using the private use area. In this case `"\u{F6FF}"`
Now when we pass the value of `$var` to the system we must not encode those private-use codepoints to UTF-8.
Instead we should reverse the private-use encoding using `decode_byte_from_char()`, which is what the `wcs2*string` family does.

Reply:
Thanks for clarifying!

Reply:
Wait, at least thus far in the rust code the only uses of `setenv_lock()`/`unsetenv_lock()` are only with fixed strings (the locale variables, the curses variables, and the timezone variable `TZ`.

Reply:
The `env_var_name.to_string()` bit is actually fine because the only uses of it are with fixed/hard-coded strings (the curses variable names, the locale variable names, and the timezone variable name `TZ`).

But the `value = var.as_string().to_string()` bit is the one that could contain user-specified content.